### PR TITLE
Convert reset_internal_energy to C++ and simplify the algorithm

### PR DESCRIPTION
**Pull request by maxpkatz**
## PR summary

We now simply evaluate the eint corresponding to small_temp, and reset eint (and eden) as needed to be at least that large. We then apply the dual energy criterion. This makes the behavior corresponding to `allow_small_energy == 0` the default, and removes that parameter.

## PR checklist

- [x] test suite needs to be run on this PR
- [x] this PR will change answers in the test suite
- [ ] all functions have docstrings as per the coding conventions
- [x] the `CHANGES` file has been updated
- [x] if appropriate, this change is described in the docs

### CODE REVIEW

**Comment by maxpkatz** [comment]
This addresses part of #702 but not all of it (reset_internal_energy is still a discrete routine).

**Comment by zingale** [comment]
Note: this seems to dramatically help the Sod stellar problem (Sod_stellar-test1-2d):

development:
Sod_stellar-test1-2d-development

this PR:
Sod_stellar-test1-2d-reset_internal_energy

File: Source/driver/Castro_setup.cpp [hunk 1]
```diff
@@ -193,6 +193,17 @@ Castro::variableSetUp ()
...
+  // with the minimum permitted by the EOS, and vice versa.
+
+  Real new_min_T = std::max(small_temp, EOSData::mintemp);
+  small_temp = new_min_T;
+  EOSData::mintemp = new_min_T;
+
+  Real new_min_rho = std::max(small_dens, EOSData::mindens);
+  small_dens = new_min_rho;
+  EOSData::mindens = new_min_rho;
+
```
Comment (zingale) [inline Source/driver/Castro_setup.cpp]: does Fortran also know about this?

**Comment by zingale** [comment]
tests are here: http://groot.astro.sunysb.edu/Castro/test-suite/gfortran/2020-03-14-006/index.html

File: Source/driver/Castro_setup.cpp (hunk 1 above)
Comment (maxpkatz) [reply Source/driver/Castro_setup.cpp]: Yes, we have the same logic in Fortran, it's just written in a slightly different way. We should have done this in the original EOS pull request (#788); that was an oversight on my part.

**Comment by maxpkatz** [comment]
I had missed that SDC called `ca_reset_internal_e` directly. So I moved the actual implementation to a separate interface that takes a box and an Array4, and then implemented both the MultiFab interface and that call from SDC with it. Please retest.

**Comment by zingale** [comment]
here are the tests: http://groot.astro.sunysb.edu/Castro/test-suite/gfortran/2020-03-15/index.html

**Pull request merged**